### PR TITLE
Use `pcg` + `randomSeed()` instead of `std::mt19937`/`std::random_device`

### PR DESCRIPTION
**Pull request by rschu1ze**
Cf. https://github.com/ClickHouse/ClickHouse/pull/54391#issuecomment-1797046449

- changes all usage of Mersenne Twister RNGs + `random_device()` by PRG RNG and `randomSeed()` 
- also extends the style check to prohibit new usages outside tests

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit e46dbcb2f1bcc26000b26a50c103547b4dab4fb2 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56430/e46dbcb2f1bcc26000b26a50c103547b4dab4fb2/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6789019484">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56430/e46dbcb2f1bcc26000b26a50c103547b4dab4fb2/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: utils/check-style/check-style [hunk 1]
```diff
@@ -16,404 +16,409 @@ LC_ALL="en_US.UTF-8"
...
+#     xargs grep $@ -P '((class|struct|namespace|enum|if|for|while|else|throw|switch).*|\)(\s*const)?(\s*override)?\s*)\{$|\s$|^ {1,3}[^\* ]\S|\t|^\s*(if|else if|if constexpr|else if constexpr|for|while|catch|switch)\(|\( [^\s\\]|\S \)' |
+# # a curly brace not in a new line, but not for the case of C++11 init or agg. initialization | trailing whitespace | number of ws not a multiple of 4, but not in the case of comment continuation | missing whitespace after for/if/while... before opening brace | whitespaces inside braces
+#     grep -v -P '(//|:\s+\*|\$\(\()| \)"'
+# # single-line comment | continuation of a multiline comment | a typical piece of embedded shell code | something like ending of raw string literal
+#
+# # Tabs
+# find $ROOT_PATH/{src,base,programs,utils} -name '*.h' -or -name '*.cpp' 2>/dev/null |
+#     grep -vP $EXCLUDE_DIRS |
+#     xargs grep $@ -F $'\t'
+#
```
Comment (alexey-milovidov) [inline utils/check-style/check-style:44]: I see all other code is commented out - it was probably committed by mistake.

Comment (rschu1ze) [reply utils/check-style/check-style:44]: Ah, I missed that, sorry. Fixed.

**Comment by evillique** [comment]
`02725_keeper_fault_inject_sequential_cleanup` fails due to `insert_keeper_fault_injection_seed` probably being no longer valid

**Comment by rschu1ze** [comment]
> `02725_keeper_fault_inject_sequential_cleanup` probably fails due to `insert_keeper_fault_injection_seed` being no longer valid

Yes, that was surely the reason. The problem did unfortunately not reproduce locally for me ...

@antonio2368 This PR replaces the usual Mersenne-Twister `std::mt19937` RNG by a faster one + it forbids new usage of `std::mt19937`. One place I fixed was "Common/ZooKeeper/ZooKeeperWithFaultInjection.h", and test "02725_keeper_fault_inject_sequential_cleanup" started to fail ([here](https://s3.amazonaws.com/clickhouse-test-reports/56430/38c76eb7098e7f63a37f10e90a6581bb182ce3ca/stateless_tests__debug__s3_storage__[2_6].html) [here](https://s3.amazonaws.com/clickhouse-test-reports/56430/38c76eb7098e7f63a37f10e90a6581bb182ce3ca/stateless_tests__release__s3_storage__[2_2].html) [here](https://s3.amazonaws.com/clickhouse-test-reports/56430/38c76eb7098e7f63a37f10e90a6581bb182ce3ca/stateless_tests__tsan__s3_storage__[2_5].html)). The test was introduced with #49418 which states `The error is only reproducible with DatabaseOrdinary and it generates only a warning in the logs (refer to the issue for more details), hence it may not be necessary to include such a test.`So the test checks a bug with ordinary databases which are obsolete by now, afaik. I therefore remove 02725_keeper_fault_inject_sequential_cleanup ([here](https://github.com/ClickHouse/ClickHouse/pull/56430/commits/504bf730b1a78f406dabc3322c787b2cd6f909a4)). Looking at the test, it appears custom-build for some bug and not testing a feature systematically. Are you okay with deleting the test?

**Review by evillique — APPROVED** [review]

**Comment by antonio2368** [comment]
@rschu1ze the test was not added by me so better to check with @vdimir

**Comment by rschu1ze** [comment]
@vdimir ^^

**Comment by vdimir** [comment]
It's a bit surprising that after changing seed test started to fail instead of just stopping reproduce some bug and checking particular code. But seems that we just do not expect `Fault injection after operation` in test. Ok, let's remove for now, I'll check if it's possible to rewrite a test.

**Comment by Algunenano** [comment]
> It's a bit surprising that after changing seed test started to fail instead of just stopping reproduce some bug and checking particular code.

The problem comes from changing the randomization. Different seed/random generation number leads to a different number of faults, which ends up reaching the 20 limit. If you want you can forget about it, I'm working on adding way more faults and tests in [https://github.com/ClickHouse/ClickHouse/pull/56389](https://github.com/ClickHouse/ClickHouse/pull/56389/files#diff-9e5613b80b115f81df382bc15be97c98c6656cc72d924fd90ce884b7010e2ee1) and that will fix the test.

**Pull request merged**